src/record_syscall.cc

```diff
@@ -3783,8 +3783,9 @@ static void process_mremap(RecordTask* t, remote_ptr<void> old_addr,
     return;
   }
 
-  // Now record the new part of the mapping.
-  km = m.map.subrange(new_addr + old_size, new_addr + new_size);
+  // Now record the new part of the mapping, possibly with some overlap.
+  km = m.map.subrange(new_addr + floor_page_size(old_length),
```

Comment:
The comment should probably say something like "Now record the new part of the mapping, possibly overlapping the last page of the previous mapping if additional bytes were mapped into that page"

The problem is slightly below here.  If we have a file with a length between pages `N` and `N+1`,  then using `old_size` (which is really `ceil_page_size(old_length)`) will give us a `KernelMapping` with `file_offset_bytes()` at `page_size * (N+1)`.  `st.st_size` is somewhere below that, so we end up with a negative value for `end` and `record_remote` explodes.

If we instead start our `KernelMapping` at `floor_page_size(old_length)` we guarantee that `end` will be positive and `record_remote` will pick up any existing bytes mapped into the page that already exists.